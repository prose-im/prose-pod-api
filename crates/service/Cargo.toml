[package]
name = "service"
version = { workspace = true }
edition = { workspace = true }
publish = false
license = { workspace = true }
authors = ["RÃ©mi Bardon <remi@remibardon.name>"]

[dependencies]
anyhow = { workspace = true }
async-trait = { workspace = true }
base64 = { workspace = true }
chrono = { workspace = true }
email_address = { workspace = true }
figment = { workspace = true }
form_urlencoded = { workspace = true }
futures = { workspace = true }
linked_hash_set = { workspace = true }
hickory-proto = { workspace = true }
hickory-resolver = { workspace = true }
hmac = { workspace = true }
iso8601-duration = { workspace = true }
jid = { workspace = true }
lazy_static = { workspace = true }
lettre = { workspace = true }
mime = { workspace = true }
minidom = { workspace = true }
parking_lot = { workspace = true }
prose-xmpp = { workspace = true }
prosody-config = { workspace = true }
rand = { workspace = true }
reqwest = { workspace = true }
sea-orm = { workspace = true }
secrecy = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = { workspace = true }
sha2 = { workspace = true }
strum = { workspace = true }
tempfile = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true }
tokio-util = { workspace = true }
tracing = { workspace = true }
url_serde = { workspace = true }
urlencoding = { workspace = true }
uuid = { workspace = true, features = ["v4"] }
xmpp-parsers = { workspace = true }

[dev-dependencies]
cucumber = { workspace = true }
insta = { workspace = true }
migration = { workspace = true }

[[test]]
name = "behavior"
harness = false
