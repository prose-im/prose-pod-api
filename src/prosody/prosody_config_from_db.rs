// prose-pod-api
//
// Copyright: 2024, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use entity::server_config::Model as ServerConfig;

use super::ProsodyConfig;

pub fn prosody_config_from_db(model: ServerConfig) -> ProsodyConfig {
    let mut config = ProsodyConfig::default();

    if model.message_archive_enabled {
        config.modules_enabled.insert("mam".to_string());
        config.archive_expires_after = Some(model.message_archive_retention);
    }

    config
}

// Some old method definitions
// FIXME: Remove useless code below

// fn add_admin(&mut self, jid: JID) {
//     self.config.admins.insert(jid);
// }
// fn remove_admin(&mut self, jid: &JID) {
//     self.config.admins.remove(jid);
// }

// fn add_enabled_module(&mut self, module_name: String) -> bool {
//     self.config.enabled_modules.insert(module_name)
// }
// fn remove_enabled_module(&mut self, module_name: &String) -> bool {
//     self.config.enabled_modules.remove(module_name)
// }

// fn add_disabled_module(&mut self, module_name: String) -> bool {
//     self.config.disabled_modules.insert(module_name)
// }
// fn remove_disabled_module(&mut self, module_name: &String) -> bool {
//     self.config.disabled_modules.remove(module_name)
// }

// fn set_rate_limit(&mut self, conn_type: ConnectionType, value: DataRate) {
//     self.config
//         .limits
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .rate = Some(value)
// }
// fn set_burst_limit(&mut self, conn_type: ConnectionType, value: DurationTime) {
//     self.config
//         .limits
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .burst = Some(value)
// }
// fn set_timeout(&mut self, value: DurationTime) {
//     self.config.limits_resolution = Some(value.seconds().clone());
// }

// fn enable_message_archiving(&mut self) -> bool {
//     self.add_enabled_module("mam".to_string())
// }
// fn disable_message_archiving(&mut self) -> bool {
//     self.remove_enabled_module(&"mam".to_string())
// }
