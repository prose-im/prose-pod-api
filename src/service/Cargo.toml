[package]
name = "service"
version = { workspace = true }
edition = { workspace = true }
publish = false
license = { workspace = true }
authors = ["RÃ©mi Bardon <remi@remibardon.name>"]

[dependencies]
anyhow = { workspace = true, features = ["backtrace"] }
async-trait = { workspace = true }
base64 = { workspace = true }
biscuit-auth = { workspace = true }
chrono = { workspace = true }
email_address = { workspace = true, features = ["serde_support"] }
figment = { workspace = true, features = ["env", "toml"] }
form_urlencoded = { workspace = true }
futures = { workspace = true }
linked_hash_set = { workspace = true, features = ["serde"] }
hickory-proto = { workspace = true, features = [
    "tokio-runtime",
    "serde-config",
] }
hickory-resolver = { workspace = true, features = ["tokio-runtime"] }
hmac = { workspace = true }
iso8601-duration = { workspace = true, features = ["serde"] }
iso8601-timestamp = { workspace = true, features = ["serde", "std"] }
jid = { workspace = true }
lazy_static = { workspace = true }
lettre = { workspace = true, features = [
    "smtp-transport",
    "rustls-tls",
    "hostname",
    "builder",
] }
mime = { workspace = true }
minidom = { workspace = true }
parking_lot = { workspace = true }
prose-xmpp = { workspace = true }
prosody-config = { workspace = true }
rand = { workspace = true }
reqwest = { workspace = true, features = [
    "blocking",
    "rustls-tls",
    "json",
    "trust-dns",
] }
sea-orm = { workspace = true, features = [
    "runtime-tokio-rustls",
    "sqlx-sqlite",
    "macros",
    "with-json",
    "with-chrono",
    "with-uuid",
] }
sea-orm-migration = { workspace = true, features = [
    "runtime-tokio-rustls",
    "sqlx-sqlite",
] }
secrecy = { workspace = true, features = ["serde"] }
semver = { workspace = true }
serde_json = { workspace = true }
serde_with = { workspace = true, features = ["macros", "alloc", "base64"] }
serdev = { workspace = true }
strum = { workspace = true, features = ["derive"] }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["rt"] }
tokio-util = { workspace = true }
tracing = { workspace = true }
unicode-normalization = { workspace = true }
url = { workspace = true }
url_serde = { workspace = true }
urlencoding = { workspace = true }
uuid = { workspace = true, features = ["v4"] }
validator = { workspace = true, features = ["derive", "unic"] }
xmpp-parsers = { workspace = true }

[dev-dependencies]
cucumber = { workspace = true, features = ["macros"] }
insta = { workspace = true }
figment = { version = "*", features = ["json"] }

[features]
default = []
test = []

[[test]]
name = "behavior"
harness = false
