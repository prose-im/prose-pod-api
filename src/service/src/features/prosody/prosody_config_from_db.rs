// prose-pod-api
//
// Copyright: 2024, Rémi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use prosody_config::{utils::def, *};
use tracing::{info, warn};

use crate::{
    app_config::{AppConfig, FILE_SHARE_HOST},
    models::{self as db},
    server_config::ServerConfig,
    ProseDefault,
};

use super::{prosody_config::util::*, prosody_overrides::ProsodyOverrides, ProsodyConfig};

pub fn prosody_config_from_db(model: ServerConfig, app_config: &AppConfig) -> ProsodyConfig {
    let mut config = prosody_config::ProsodyConfig::prose_default(&model, app_config);

    let global_settings = &mut config.global_settings;
    let muc_settings = config
        .additional_sections
        .iter_mut()
        .find_map(|section| match section {
            ProsodyConfigSection::Component {
                plugin, settings, ..
            } if plugin.as_str() == "muc" => Some(settings),
            _ => None,
        })
        .expect("The 'Chatrooms' section should always be present.");

    if model.message_archive_enabled {
        add_enabled_module(global_settings, "mam");
        global_settings.archive_expires_after =
            Some(model.message_archive_retention.into_prosody());
        global_settings.default_archive_policy = Some(ArchivePolicy::Always);
        global_settings.max_archive_query_results = Some(100);
        add_enabled_module(muc_settings, "muc_mam");
    }

    if model.file_upload_allowed {
        config
            .additional_sections
            .push(ProsodyConfigSection::Component {
                hostname: FILE_SHARE_HOST.into(),
                plugin: "http_file_share".into(),
                name: "HTTP File Upload".into(),
                settings: ProsodySettings {
                    http_file_share_size_limit: Some(Bytes::MebiBytes(20)),
                    http_file_share_daily_quota: Some(Bytes::MebiBytes(250)),
                    http_file_share_expires_after: Some(
                        model.file_storage_retention.into_prosody(),
                    ),
                    http_host: Some("localhost".into()),
                    http_external_url: Some("http://localhost:5280".into()),
                    ..Default::default()
                },
            })
    }

    if model.federation_enabled {
        global_settings.unmark_disabled("s2s");
        global_settings.enable_module("s2s_bidi".to_owned());
        global_settings.s2s_ports = Some(vec![5269]);
        global_settings.s2s_require_encryption = Some(true);
        global_settings.s2s_secure_auth = Some(false);
        global_settings.limits.get_or_insert_default().insert(
            ConnectionType::ServerToServerInbounds,
            ConnectionLimits {
                rate: Some(DataRate::KiloBytesPerSec(250)),
                burst: Some(Duration(TimeLike::Seconds(4))),
            },
        );

        if model.federation_whitelist_enabled {
            global_settings.enable_module("s2s_whitelist".to_owned());
            global_settings.s2s_whitelist = Some(model.federation_friendly_servers);
        }
    }

    if model.c2s_unencrypted {
        warn!("Debug config `c2s_unencrypted` is enabled.");
        global_settings.enable_module("reload_modules".to_owned());
        global_settings.custom_settings.push(Group::new(
            "Debug config: c2s_unencrypted",
            vec![
                def("c2s_require_encryption", false),
                def("allow_unencrypted_plain_auth", true),
                def("reload_modules", vec!["saslauth"]),
            ],
        ));
    }

    for module in app_config.prosody.additional_modules_enabled.iter() {
        config.global_settings.enable_module(module.clone());
    }

    if let Some(overrides) = model.prosody_overrides {
        match serde_json::from_value::<ProsodyOverrides>(overrides) {
            Ok(ProsodyOverrides {
                c2s_require_encryption,
            }) => {
                info!("Applying overrides to the generated Prosody configuration file…");
                macro_rules! override_global_setting {
                    ($var:ident) => {
                        if let Some($var) = $var {
                            config.global_settings.$var = Some($var);
                        }
                    };
                }
                override_global_setting!(c2s_require_encryption);
            },
            Err(err) => warn!("Prosody overrides stored in database cannot be read, they won’t be applied. To fix this, call `PUT /v1/server/config/prosody-overrides` with a new value. You can `GET /v1/server/config/prosody-overrides` first to see what the stored value was. Error: {err}"),
        }
    }

    ProsodyConfig(config)
}

// ===== Model mappings =====

pub trait IntoProsody<T> {
    fn into_prosody(self) -> T;
}

impl IntoProsody<DateLike> for db::DateLike {
    fn into_prosody(self) -> DateLike {
        match self {
            Self::Days(n) => DateLike::Days(n),
            Self::Weeks(n) => DateLike::Weeks(n),
            Self::Months(n) => DateLike::Months(n),
            Self::Years(n) => DateLike::Years(n),
        }
    }
}

impl IntoProsody<TimeLike> for db::TimeLike {
    fn into_prosody(self) -> TimeLike {
        match self {
            Self::Seconds(n) => TimeLike::Seconds(n),
            Self::Minutes(n) => TimeLike::Minutes(n),
            Self::Hours(n) => TimeLike::Hours(n),
        }
    }
}

impl<A, B> IntoProsody<Duration<B>> for db::Duration<A>
where
    A: db::DurationContent + IntoProsody<B>,
    B: DurationContent,
{
    fn into_prosody(self) -> Duration<B> {
        Duration(self.0.into_prosody())
    }
}

impl<A, B> IntoProsody<PossiblyInfinite<B>> for db::PossiblyInfinite<A>
where
    A: IntoProsody<B>,
{
    fn into_prosody(self) -> PossiblyInfinite<B> {
        match self {
            Self::Infinite => PossiblyInfinite::Infinite,
            Self::Finite(d) => PossiblyInfinite::Finite(d.into_prosody()),
        }
    }
}

impl IntoProsody<ConnectionType> for db::xmpp::XmppDirectionalConnectionType {
    fn into_prosody(self) -> ConnectionType {
        match self {
            Self::ClientToServer => ConnectionType::ClientToServer,
            Self::ServerToServerInbounds => ConnectionType::ServerToServerInbounds,
            Self::ServerToServerOutbounds => ConnectionType::ServerToServerOutbounds,
        }
    }
}
