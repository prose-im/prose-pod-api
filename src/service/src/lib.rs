// prose-pod-api
//
// Copyright: 2023–2024, Rémi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

extern crate biscuit_auth as biscuit;
pub extern crate prose_xmpp;
pub extern crate prosody_config;
pub extern crate reqwest;
extern crate xmpp_parsers;

pub mod dependencies;
pub mod errors;
mod features;
mod migrations;
pub mod models;
mod mutation_error;
pub mod util;

pub use features::app_config::AppConfig;
pub use features::prosody;
pub use features::server_config::ServerConfig;
pub use features::*;
pub use linked_hash_set::LinkedHashSet;
pub use migrations::Migrator;
pub use mutation_error::MutationError;
pub use prosody::prosody_config_from_db;
pub use prosody_config::ProsodyConfigSection;
pub use reqwest::Client as HttpClient;
pub use sea_orm;
pub use sea_orm_migration::MigratorTrait;

trait ProseDefault {
    fn prose_default(server_config: &ServerConfig, app_config: &AppConfig) -> Self;
}
