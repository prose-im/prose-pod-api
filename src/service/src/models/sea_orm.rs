// prose-pod-api
//
// Copyright: 2025, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use std::{
    fmt::{Debug, Display},
    ops::Deref,
    str::FromStr,
};

// ===== AsString =====

pub struct SeaOrmAsString<T: Display + FromStr>(pub T);

impl<T: Display + FromStr> Display for SeaOrmAsString<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.0, f)
    }
}
impl<T: Display + FromStr> FromStr for SeaOrmAsString<T> {
    type Err = <T as FromStr>::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        T::from_str(s).map(Self)
    }
}
impl<T: Display + FromStr + Debug> Debug for SeaOrmAsString<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(&self.0, f)
    }
}
impl<T: Display + FromStr + Clone> Clone for SeaOrmAsString<T> {
    fn clone(&self) -> Self {
        Self(self.0.clone())
    }
}
impl<T: Display + FromStr + PartialEq> PartialEq for SeaOrmAsString<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0.eq(&other.0)
    }
}
impl<T: Display + FromStr> From<T> for SeaOrmAsString<T> {
    fn from(value: T) -> Self {
        Self(value)
    }
}

impl<T: Display + FromStr> Deref for SeaOrmAsString<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: Display + FromStr> From<SeaOrmAsString<T>> for sea_orm::sea_query::Value {
    crate::sea_orm_string_value_impl!(SeaOrmAsString<T>);
}
impl<T: Display + FromStr> sea_orm::TryGetable for SeaOrmAsString<T>
where
    <T as FromStr>::Err: Debug,
{
    crate::sea_orm_try_get_by_string!(using: FromStr);
}
impl<T: Display + FromStr> sea_orm::sea_query::ValueType for SeaOrmAsString<T> {
    crate::sea_orm_string_value_type_impl!(SeaOrmAsString<T>, None, using: FromStr);
}
impl<T: Display + FromStr> sea_orm::sea_query::Nullable for SeaOrmAsString<T> {
    crate::sea_orm_string_nullable_impl!();
}

// ===== LinkedStringSet =====

#[derive(Clone, Debug, PartialEq, Eq)]
#[derive(serde::Serialize, serde::Deserialize)]
#[derive(sea_orm::FromJsonQueryResult)]
#[repr(transparent)]
pub struct LinkedStringSet(pub linked_hash_set::LinkedHashSet<String>);

impl std::ops::Deref for LinkedStringSet {
    type Target = linked_hash_set::LinkedHashSet<String>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl From<linked_hash_set::LinkedHashSet<String>> for LinkedStringSet {
    fn from(value: linked_hash_set::LinkedHashSet<String>) -> Self {
        Self(value)
    }
}

impl std::fmt::Display for LinkedStringSet {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{{")?;
        let len = self.len();
        for (i, s) in self.0.iter().enumerate() {
            write!(f, "{s:?}")?;
            if i < len - 1 {
                write!(f, ", ")?;
            }
        }
        write!(f, "}}")
    }
}
