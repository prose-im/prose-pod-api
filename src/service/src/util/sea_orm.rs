// prose-pod-api
//
// Copyright: 2024–2025, Rémi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

#[doc(hidden)]
#[macro_export]
macro_rules! sea_orm_try_get_by_string {
    (using: FromStr) => {
        fn try_get_by<I: sea_orm::ColIdx>(
            res: &sea_orm::QueryResult,
            index: I,
        ) -> Result<Self, sea_orm::TryGetError> {
            // https://github.com/SeaQL/sea-orm/discussions/1176#discussioncomment-4024088
            let value = res
                .try_get_by(index)
                .map_err(sea_orm::TryGetError::DbErr)
                .and_then(|opt: Option<String>| {
                    opt.ok_or(sea_orm::TryGetError::Null(format!("{index:?}")))
                })?;
            <Self as std::str::FromStr>::from_str(value.as_str())
                // Technically, the value is not `null`, but we wouldn't want to unsafely unwrap here.
                .map_err(|e| sea_orm::TryGetError::Null(format!("{:?}", e)))
        }
    };
    (using: From<String>) => {
        fn try_get_by<I: sea_orm::ColIdx>(
            res: &sea_orm::QueryResult,
            index: I,
        ) -> Result<Self, sea_orm::TryGetError> {
            // https://github.com/SeaQL/sea-orm/discussions/1176#discussioncomment-4024088
            (res.try_get_by(index))
                .map_err(sea_orm::TryGetError::DbErr)
                .map(<Self as From<String>>::from)
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! sea_orm_string_value_impl {
    ($t:ty) => {
        fn from(value: $t) -> Self {
            Self::String(Some(Box::new(value.to_string())))
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! sea_orm_string_value_type_impl {
    ($t:ty, $length:expr, using: FromStr) => {
        fn try_from(
            v: sea_orm::sea_query::Value,
        ) -> Result<Self, sea_orm::sea_query::ValueTypeErr> {
            match v {
                sea_orm::sea_query::Value::String(Some(value)) => {
                    <Self as std::str::FromStr>::from_str(value.as_str())
                        .map_err(|_| sea_orm::sea_query::ValueTypeErr)
                }
                _ => Err(sea_orm::sea_query::ValueTypeErr),
            }
        }

        fn type_name() -> String {
            stringify!($t).to_string()
        }

        fn array_type() -> sea_orm::sea_query::ArrayType {
            sea_orm::sea_query::ArrayType::String
        }

        fn column_type() -> sea_orm::sea_query::ColumnType {
            sea_orm::sea_query::ColumnType::string($length)
        }
    };
    ($t:ty, $length:expr, using: From<String>) => {
        fn try_from(
            v: sea_orm::sea_query::Value,
        ) -> Result<Self, sea_orm::sea_query::ValueTypeErr> {
            match v {
                sea_orm::sea_query::Value::String(Some(value)) => {
                    Ok(<Self as From<String>>::from(*value))
                }
                _ => Err(sea_orm::sea_query::ValueTypeErr),
            }
        }

        fn type_name() -> String {
            stringify!($t).to_string()
        }

        fn array_type() -> sea_orm::sea_query::ArrayType {
            sea_orm::sea_query::ArrayType::String
        }

        fn column_type() -> sea_orm::sea_query::ColumnType {
            sea_orm::sea_query::ColumnType::string($length)
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! sea_orm_string_nullable_impl {
    () => {
        fn null() -> sea_orm::Value {
            sea_orm::Value::String(None)
        }
    };
}

#[macro_export]
macro_rules! sea_orm_string {
    ($t:ty) => {
        crate::sea_orm_string!($t, length: None);
    };
    ($t:ty, length: $length:expr) => {
        impl From<$t> for sea_orm::sea_query::Value {
            crate::sea_orm_string_value_impl!($t);
        }

        impl sea_orm::TryGetable for $t {
            crate::sea_orm_try_get_by_string!(using: FromStr);
        }

        impl sea_orm::sea_query::ValueType for $t {
            crate::sea_orm_string_value_type_impl!($t, $length, using: FromStr);
        }

        impl sea_orm::sea_query::Nullable for $t {
            crate::sea_orm_string_nullable_impl!();
        }
    };
    ($t:ty; enum) => {
        crate::sea_orm_string!(
            $t,
            length: Some(
                <Self as strum::IntoEnumIterator>::iter()
                    .map(|v| v.to_string().len())
                    .max()
                    .unwrap() as u32
            )
        );
    };
    ($t:ty; secret) => {
        impl From<$t> for sea_orm::sea_query::Value {
            fn from(value: $t) -> Self {
                use secrecy::ExposeSecret;
                Self::String(Some(Box::new(value.expose_secret().to_string())))
            }
        }

        impl sea_orm::TryGetable for $t {
            crate::sea_orm_try_get_by_string!(using: From<String>);
        }

        impl sea_orm::sea_query::ValueType for $t {
            crate::sea_orm_string_value_type_impl!($t, None, using: From<String>);
        }

        impl sea_orm::sea_query::Nullable for $t {
            crate::sea_orm_string_nullable_impl!();
        }
    };
}
