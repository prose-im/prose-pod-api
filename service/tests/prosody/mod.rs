// prose-pod-api
//
// Copyright: 2024, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use ::service::prosody_config_from_db;
use cucumber::{given, then, when};
use insta::assert_snapshot;
use sea_orm::*;

use crate::TestWorld;

#[given("nothing has changed since the initialization of the workspace")]
fn given_nothing_changed(_world: &mut TestWorld) {
    // Do nothing, even though we could performs checks
}

#[given("every optional feature has been disabled")]
async fn given_everything_disabled(world: &mut TestWorld) -> Result<(), DbErr> {
    let mut model = world.server_config.clone().into_active_model();

    model.message_archive_enabled = Set(false);
    model.file_upload_allowed = Set(false);
    model.mfa_required = Set(false);
    model.federation_enabled = Set(false);

    world.server_config = model.update(&world.db).await?;
    Ok(())
}

#[when("generating a new Prosody configuration file from the database")]
fn when_generating_prosody_config(world: &mut TestWorld) {
    let prosody_config = prosody_config_from_db(world.server_config.clone());
    world.prosody_config = Some(prosody_config);
}

#[then(expr = "the file should match the snapshot named {string}")]
fn then_prosody_config_file_matches(world: &mut TestWorld, snapshot_name: String) {
    assert_snapshot!(snapshot_name, world.prosody_config().to_string())
}
