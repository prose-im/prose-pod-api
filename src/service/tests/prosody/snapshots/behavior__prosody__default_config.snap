---
source: service/tests/prosody/mod.rs
assertion_line: 43
expression: world.prosody_config().to_string()
---
-- Prose Pod Server
-- XMPP Server Configuration
-- /!\ This file has been automatically generated by Prose Pod API.
-- /!\ Do NOT edit this file manually or your changes will be overridden during the next reload.

-- Base server configuration
pidfile = "/var/run/prosody/prosody.pid"

authentication = "internal_hashed"

default_storage = "internal"

log = {
  info = "*console";
}

-- Network interfaces/ports
http_ports = { 5280 }

-- Modules
plugin_paths = { "/usr/local/lib/prosody/modules" }
modules_enabled = {
  "auto_activate_hosts";
  "roster";
  "groups_internal";
  "saslauth";
  "tls";
  "dialback";
  "disco";
  "posix";
  "smacks";
  "private";
  "vcard_legacy";
  "vcard4";
  "version";
  "uptime";
  "time";
  "ping";
  "lastactivity";
  "pep";
  "blocklist";
  "limits";
  "carbons";
  "csi";
  "server_contact_info";
  "websocket";
  "reload_modules";
  "cloud_notify";
  "register";
  "prose_version";
  "mam";
}
modules_disabled = { "s2s" }

-- Disable in-band registrations (done through the Prose Pod Dashboard/API)
allow_registration = false

-- Mandate highest security levels
c2s_require_encryption = true

-- Enforce safety C2S/S2S limits
c2s_stanza_size_limit = 256 * 1024

limits = {
  c2s = {
    rate = "50kb/s";
    burst = "2s";
  };
}

-- Allow reverse-proxying to WebSocket service over insecure local HTTP
consider_websocket_secure = true

-- MAM settings
archive_expires_after = "never"
default_archive_policy = true
max_archive_query_results = 100

-- Enable vCard legacy compatibility layer
upgrade_legacy_vcards = true

-- mod_reload_modules
reload_modules = { "tls" }

-- mod_cloud_notify
push_notification_with_body = true
push_notification_with_sender = true

-- Server hosts and components
VirtualHost "test.org"
  admins = { "prose-pod-api@admin.prose.local" }

  -- Modules
  modules_enabled = {
    "rest";
    "http_oauth2";
    "admin_rest";
  }

  -- HTTP settings
  http_host = "prose-pod-server"

  -- XEP-0030: Service Discovery
  disco_items = {
    {
      "upload.prose.local";
      "HTTP File Upload";
    };
  }

  -- mod_http_oauth2
  allowed_oauth2_grant_types = {
    "authorization_code";
    "refresh_token";
    "password";
  }
  oauth2_access_token_ttl = 10800
  oauth2_refresh_token_ttl = 0
  oauth2_registration_key = "dummy-test-key"

VirtualHost "admin.prose.local"
  admins = { "prose-pod-api@admin.prose.local" }

  -- Modules
  modules_enabled = {
    "admin_rest";
    "init_admin";
  }

  -- HTTP settings
  http_host = "prose-pod-server-admin"

Component "groups.test.org" "muc"
  name = "Chatrooms"

  -- Modules
  modules_enabled = {
    "muc_public_affiliations";
    "muc_mam";
  }

  -- MAM settings
  max_archive_query_results = 100

  restrict_room_creation = "local"

  -- MUC settings
  muc_log_all_rooms = true
  muc_log_by_default = true
  muc_log_expires_after = "never"

Component "upload.prose.local" "http_file_share"
  name = "HTTP File Upload"

  -- HTTP settings
  http_external_url = "https://prose.test.org/"
  http_paths = {
    file_share = "/upload";
  }

  -- XEP-0363: HTTP File Upload
  http_file_share_size_limit = 20 * 1024 * 1024
  http_file_share_expires_after = "never"
  http_file_share_daily_quota = 100 * 1024 * 1024
  http_file_share_access = { "test.org" }
